src/BalanceTx/Collateral.purs

```diff
@@ -0,0 +1,252 @@
+      collReturnAda :: BigInt
+      collReturnAda = unwrap $
+        Max (collAdaValue - minRequiredCollateral) <> Max minAdaValue
+
+      -- Build the final collateral return output:
+      collReturnOutput :: TransactionOutput
+      collReturnOutput = wrap $
+        collReturnOutputRec
+          { amount = Value.mkValue (wrap collReturnAda) collNonAdaAsset }
+
+      totalCollateral :: BigInt
+      totalCollateral = collAdaValue - collReturnAda
+
+    except $
+      case totalCollateral > zero of
+        true ->
+          -- Set collateral return and total collateral:
+          Right $
+            transaction # _body <<< _collateralReturn ?~ collReturnOutput
+              # _body <<< _totalCollateral ?~ wrap totalCollateral
+        false ->
+          Left $ CollateralReturnError
+            "Negative totalCollateral after covering min-utxo-ada requirement."
+
+--------------------------------------------------------------------------------
+-- Select Collateral
+--------------------------------------------------------------------------------
+
+collateralReturnMinAdaValue
+  :: List TransactionUnspentOutput -> QueryM (Maybe BigInt)
+collateralReturnMinAdaValue =
+  utxoMinAdaValue <<< fakeOutputWithNonAdaAssets <<< foldMap nonAdaAsset
+
+type ReturnOutMinAdaValue = BigInt
+
+newtype CollateralCandidate =
+  CollateralCandidate (List TransactionUnspentOutput /\ ReturnOutMinAdaValue)
+
+derive instance Newtype CollateralCandidate _
+
+instance Eq CollateralCandidate where
+  eq = eq `on` (Tuple.snd <<< unwrap)
+
+instance Ord CollateralCandidate where
+  compare lhs rhs =
+    case on compare (Tuple.snd <<< unwrap) lhs rhs of
+      -- If two candidate utxo combinations correspond to return outputs with
+      -- the same utxo min ada value, order them by the number of
+      -- collateral inputs:
+      EQ -> on compare (List.length <<< Tuple.fst <<< unwrap) lhs rhs
+      ordering -> ordering
+
+mkCollateralCandidate
+  :: List TransactionUnspentOutput /\ Maybe ReturnOutMinAdaValue
+  -> Maybe CollateralCandidate
+mkCollateralCandidate (unspentOutputs /\ returnOutMinAdaValue) =
+  CollateralCandidate <<< Tuple unspentOutputs <$> returnOutMinAdaValue
+
+-- | Selects an utxo combination to use as collateral by generating all possible
+-- | utxo combinations and then applying the following constraints:
+-- |
+-- |   1. `maxCollateralInputs` protocol parameter limits the maximum
+-- |   cardinality of a single utxo combination.
+-- |
+-- |   2. Collateral inputs must have a total value of at least 5 Ada
+-- |   (`minRequiredCollateral`).
+-- |
+-- |   3. We prefer utxo combinations that require the lowest utxo min ada
+-- |   value for the corresponding collateral output, thus maintaining a
+-- |   sufficient `totalCollateral`.
+-- |
+-- |   4. If two utxo combinations correspond to return outputs with the same
+-- |   utxo min ada value, we prefer the one with fewer inputs.
+-- |
+selectCollateral
```

Comment:
My only concern here is time complexity. It's not uncommon to have hunderds of UTxOs. Could you please check the algorithm by faking the inputs (provide a list with increasing length). We need to know how many UTxOs can we handle before the performance becomes a problem (say before we get freeze time in range of seconds).

Reply:
@klntsky Yes, that's a valid point. My initial assumption was that we could sacrifice a bit of performance in favor of correctness. I'll add unit tests for the algorithm then, and think towards possible optimization options if performance becomes an issue.